Firmware/meatpack.cpp

```diff
@@ -0,0 +1,298 @@
+// NOT storing this with PROGMEM, given how frequently this table will be accessed.
+uint8_t MeatPackLookupTbl[16] = {
+    '0',	// 0000
+    '1',	// 0001
+    '2',	// 0010
+    '3',	// 0011
+    '4',	// 0100
+    '5',	// 0101
+    '6',	// 0110
+    '7',	// 0111
+    '8',	// 1000
+    '9',	// 1001
+    '.',	// 1010
+    ' ',	// 1011
+    '\n',	// 1100
+    'G',	// 1101
+    'X',	// 1110
+    '\0' // never used, 0b1111 is used to indicate next 8-bits is a full character
+};
+#else
+inline uint8_t get_char(register uint8_t in) {
+    switch (in) {
+    case 0b0000:
+        return '0';
+        break;
+    case 0b0001:
+        return '1';
+        break;
+    case 0b0010:
+        return '2';
+        break;
+    case 0b0011:
+        return '3';
+        break;
+    case 0b0100:
+        return '4';
+        break;
+    case 0b0101:
+        return '5';
+        break;
+    case 0b0110:
+        return '6';
+        break;
+    case 0b0111:
+        return '7';
+        break;
+    case 0b1000:
+        return '8';
+        break;
+    case 0b1001:
+        return '9';
+        break;
+    case 0b1010:
+        return '.';
+        break;
+    case 0b1011:
+        return ' ';
+        break;
+    case 0b1100:
+        return '\n';
+        break;
+    case 0b1101:
+        return 'G';
+        break;
+    case 0b1110:
+        return 'X';
+        break;
+    }
+    return 0;
+}
+#endif
+
+
+// State variables
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+volatile uint8_t mp_active = 0;             // Defines if MeatPack is active
+volatile uint8_t mp_cmd_active = 0;         // Is a command is pending
+volatile uint8_t mp_char_buf = 0;           // Buffers a character if dealing with out-of-sequence pairs
+volatile uint8_t mp_cmd_count = 0;          // Counts how many command bytes are received (need 2)
+volatile uint8_t mp_full_char_queue = 0;       // Counts how many full-width characters are to be received
+volatile uint8_t mp_char_out_buf[2];
+volatile uint8_t mp_char_out_count = 0;
+
+// #DEBUGGING
+#ifdef MP_DEBUG
+volatile uint32_t mp_chars_decoded = 0;
+#endif
+
+inline void mp_handle_output_char(const uint8_t c) {
+    mp_char_out_buf[mp_char_out_count++] = c;
+
+#ifdef MP_DEBUG
+    if (mp_chars_decoded < 64) {
+        ++mp_chars_decoded;
+        SERIAL_ECHOPGM("Rec Byte: ");
+        MYSERIAL.print("0x");
+        MYSERIAL.print((uint8_t)c, HEX);
+        SERIAL_ECHOLNPGM("");
+    }
+#endif
+}
+
+// Storing
+// packed = ((low & 0xF) << 4) | (high & 0xF);
+
+// Unpacking
+// low = (packed >> 4) & 0xF;
+// high = (packed & 0xF);
+
+//==========================================================================
+inline uint8_t mp_unpack_chars(const uint8_t pk, uint8_t* __restrict const chars_out) {
+    register uint8_t out = 0;
+
+#ifdef USE_LOOKUP_TABLE
+    // If lower 4 bytes is 0b1111, the higher 4 are unused, and next char is full.
```

Comment:
did you do some comparison in terms of performance gain of either of these methods? Lookup table vs. direct computation?

Please note the AVR cannot shift by 4 -> these will be 4 separate instructions

Reply:
I loaded up both methods in both Ghidra and IDA, and the disassembly was byte-for-byte identical. 

The compiler optimizes part of the array away completely anyway, and figured out some math to compute an output byte. For the switch statement, it ends up substituting the case jumps for a direct memory access.